### PR TITLE
AF-2577: External Components Support

### PR DESCRIPTION
**Pull request by jesuino**
Support for external components in Dashbuilder
--

With this PR we are bringing support for external components in Dashbuilder. Any kind of technology that can run in a browser can be used in dashbuilder without having to build the webapp or do Java programming.

### Configuration

External components can be placed on a directory configured by system property **dashbuilder.components.dir**, the default value is **/tmp/dashbuilder/components**. 

It is also possible to completely disable components using the system property **dashbuilder.components.enable**. Setting it as "false" will disable external components.

### How does it work?

We plug a view in dashbuilder that will lookup components from a standard location **{base path}/dashbuilder/components/{componentId}/index.html** so we have two main pieces:

* View: The main view is accessed using ExternalComponentPresenter and currently this presenter is directly used in a page or it is wrapped in a dashbuilder displayer when the component needs to consume data from Dashbuilder (dataset);
* Server: The servlet ExternalComponentServlet provide access for the filesystem where components are installed. ExternalComponentPresenter lookup for components through this servlet.

Another important part is about how we do communicate with components. Right now we created Javascript objects that can be sent across the component's iframe using *window.postMessage*. Check `ExternalComponentMessage`.  Later we plan to create a formal and well documented API.

Finally a descriptor `manifest.json` will be required to register the component and provide more information.

### Hello World component 

To have a better understand of how it works let's create a very simple component:

* Have Business Central running;
* In the dir configured for components create a new directory - this directory name is the component id, so let's say it is hello;
* Inside hello create a file manifest.json with the following content:

~~~
{
    "name": "Hello World",
    "parameters": []
}
~~~

* Create another file `index.html` with any HTML content you want:

~~~
<h2>Hello World!</h2>
~~~

* Now go to Business Central, Page designer and see your component under "External Components" section:

image

* Drag it to the page and you should see your HTML being rendered.

Notes:

* Live reload is support, so just develop your component and switch between Preview and Editor and you should see your changes;
* All your available components is exported with the Data Transfer tool, Runtime should support importing new components as well, but in Runtime we have a cache configured by default, so live reload is not supported there.

### Other components tasks on this PR:

This PR actually includes the following tasks:

AF-2577 | Define API for plugging External Components into Dashbuilder
AF-2578 | Support External Component Parameters 

### More components

You can see some components available in this repo: https://github.com/jesuino/dashbuilder-components

Other components in action can be seen in the following GIF:

evaluation_utilization_clean_db

Thanks for review and let me know if you face any issue!

Part of an ensemble:

https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1419
https://github.com/kiegroup/kie-wb-distributions/pull/1048

### CODE REVIEW

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
jenkins retest this

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins retest this please

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
jenkins retest this

**Comment by caponetto** [comment]
Hi @jesuino, can you please take a look at these issues that SonarCloud has raised?

**Comment by jesuino** [comment]
jenkins retest this

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Hello @caponetto I made most of the Sonar Cloud changes, some are out of my control, like do not use a deprecated class - because I need to component stuff into displayer API, and some parts of it use legacy API.

**Comment by sthundat** [comment]
Jenkins execute full downstream build

File: dashbuilder/dashbuilder-client/dashbuilder-displayer-editor/src/main/java/org/dashbuilder/client/editor/external/ExternalComponentGroupProvider.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    @AfterInitialization
+    public void loadComponents() {
+        externalComponentService.call((List<ExternalComponent> components) -> loadedComponents = components)
+                                .listComponents();
+    }
+
+    @Override
+    public String getName() {
+        return "External Components";
```
Comment (caponetto) [inline dashbuilder/dashbuilder-client/dashbuilder-displayer-editor/src/main/java/org/dashbuilder/client/editor/external/ExternalComponentGroupProvider.java:54]: Consider using `i18n` instead of hardcoding the string `External Components`.

File: dashbuilder/dashbuilder-client/dashbuilder-displayer-editor/src/main/java/org/dashbuilder/client/editor/external/ExternalComponentGroupProvider.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    Caller<ExternalComponentService> externalComponentService;
+
+    @Inject
+    SyncBeanManager beanManager;
+
+    List<ExternalComponent> loadedComponents = Collections.emptyList();
+
+    @AfterInitialization
+    public void loadComponents() {
+        externalComponentService.call((List<ExternalComponent> components) -> loadedComponents = components)
```
Comment (caponetto) [inline dashbuilder/dashbuilder-client/dashbuilder-displayer-editor/src/main/java/org/dashbuilder/client/editor/external/ExternalComponentGroupProvider.java:48]: Do we need a loading spinner here?

File: dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/ExternalComponentEditorView.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.componentEditorRoot
+-->
+<div class="componentEditorRootContainer">
+    <div class="container">
+        <div class="row">
+            <div class="col-md-3 componentEditorPropertiesColumn"></div>
+            <div class="col-md-9 componentEditorPreviewColumn"></div>
```
Comment (caponetto) [inline dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/ExternalComponentEditorView.html:20]:
```suggestion
            <div class="col-md-3" data-field="componentEditorPropertiesColumn"></div>
            <div class="col-md-9" data-field="componentEditorPreviewColumn"></div>
```

**Comment by sthundat** [comment]
@jesuino ,
I am progressing with the verification of this PR.As we discussed,found an issue if we disable dashbuilder components(dashbuilder.components.enable=false).In that case we can remove "External Component" drag group from the layout editor.currently this is displaying in page perspective just for nothing.I can create a Jira for this and  can be fix later as some changes involved in this.
Basically functionality is working fine in BC and dashbuilder runtime but I need to verify with some more sample components ,then I can approve the PR .
Thanks,
Savitha

**Comment by jesuino** [comment]
Had to update the version for the server side components project

**Comment by jesuino** [comment]
Jenkins execute full downstream build

File: dashbuilder/dashbuilder-backend/dashbuilder-services/src/main/java/org/dashbuilder/transfer/DataTransferServicesImpl.java [hunk 4]
```diff
@@ -146,6 +150,14 @@ public String doExport(DataTransferExportModel exportModel) throws java.io.IOExc
             zipFileSystem(navigationFS, zos, readmeFilter);
         }
 
+        String componentsPath = externalComponentLoader.getExternalComponentsDir();
+
+        if (componentsPath != null) {
+            zipComponents(Paths.get("file:" + componentsPath),
```
[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-backend/dashbuilder-services/src/main/java/org/dashbuilder/transfer/DataTransferServicesImpl.java:156]: it works but I guess it only works because I am on Linux (???)

File: dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/ExternalComponentEditorView.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    @DataField
+    HTMLDivElement componentEditorPreviewColumn;
+
+    @Inject
+    Elemental2DomUtil domUtil;
+
+    @Override
+    public void init(ExternalComponentEditor presenter) {
+        componentEditorPropertiesColumn.innerHTML = "";
+        componentEditorPreviewColumn.innerHTML = "";
```
[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/ExternalComponentEditorView.java:49]: in this specific case I may externally call init again and it requires a cleanup

[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/ExternalComponentEditorView.java:49]: I see now, I made confusion with other init :-D

File: dashbuilder/dashbuilder-client/dashbuilder-displayer-editor/src/main/java/org/dashbuilder/client/editor/external/ExternalComponentGroupProvider.java (hunk 2 above)
Comment (jesuino) [reply dashbuilder/dashbuilder-client/dashbuilder-displayer-editor/src/main/java/org/dashbuilder/client/editor/external/ExternalComponentGroupProvider.java:48]: this should be loaded in background before user open layout editor itself hence why I didn't add it

**Comment by jesuino** [comment]
Hello @caponetto and @sthundat   (cc @barboras7 )

I made most of the Guilherme's review - thank you so much, Guilherme!

I also made the change suggested by Savitha, it required a change at the Layout Component level, which I usually avoid modifying because it impacts all the other tools built on top of it (e.g. forms) - but I made a "safe fix" which just check null and return true by default, see:

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Hello @caponetto  - thanks for noticing the test failure! I made the suggested changes

**Comment by barboras7** [comment]
jenkins execute fdb

**Comment by mareknovotny** [comment]
Jenkins do fdb

**Comment by barboras7** [comment]
Thanks for your answers, @jesuino.

1. Thanks for taking care of this.
2. You are right, the refreshing helped. The only issue is that this might not be so obvious that the refreshing helps as for example, navigating from pages perspective to some other part of the business central and then going back to pages does not help.
3. Okay, I understand the reasons.
4. Thanks for explanation.

**Review by barboras7 — APPROVED** [review]

**Comment by mareknovotny** [comment]
Jenkins do cdb

**Comment by jesuino** [comment]
Failed tests are not related:

~~~
Test Result (5 failures / +1)
org.kie.server.integrationtests.common.rest.RestMalformedRequestIntegrationTest.testCreateContainerNonExistingGAV2[0: JAXB KieServicesConfiguration{transport=REST, serverUrl='http://localhost:43911/kie-server-services/services/rest/server'}]
org.kie.karaf.itest.KieSpringjBPMPersistenceKarafIntegrationTest.testJbpmRuntimeManagerWithPersistence
org.kie.karaf.itest.KieSpringjBPMPersistenceKarafIntegrationTest.testJbpmRuntimeManager
org.kie.server.springboot.samples.KieServerTest.testProcessStartAndWorkOnUserTask
org.kie.server.springboot.samples.KieServerMigrationTest.testProcessStartAndAbort
~~~

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute cdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
jenkins execute fdb

**Comment by jesuino** [comment]
There are three unrelated failures:

```
org.kie.hacep.PodAsReplicaTest.processOneSentMessageAsLeaderAndThenReplicaTest
org.kie.server.springboot.samples.KieServerTest.testProcessStartAndWorkOnUserTask
org.kie.server.springboot.samples.KieServerMigrationTest.testProcessStartAndAbort
```

IMO this should be merged!
